### DISCUSSION TITLE
Overview of Ideas for Apps

### DISCUSSION THREAD

**Discussion by mclaeysb** [allmaps/allmaps#103] [Ideas for Apps: Editor, Viewer, ...]
Here's an overview of ideas and feature requests for the Allmaps apps (Editor, Viewer, ...) sourcing from the corresponding discussions category and related issues.

# Annotation

- Implement georeference Extension spec fully (including canvas elements, embedding, ...)

# Editor

- General
  - Custom Background. Taken up in #100 
  - Bring Editor to the monorepo. Taken up in #93 
  - Manage annotations referencing localhost IIIF resources: don't store those and warn that these must be stored locally
  - Support [navPlace extension](https://iiif.io/api/extension/navplace/): when starting from a manifest with a navPlace element, center the map at that location
- Improve editing flow
  - If only one map is found, skip the 'Collection' step
  - Improve 'Georeference' step, collecting GCPs
    - Improve how GCPs are collected (see below).
    - Allow selecting a transformation type from a dropdown menu
    - Allow specifying a projection (see relevant discussion)
  - Place the 'Mask' step after the 'Georeference' step such that information from that step can be used.
  - Improve 'Results' step
    - Consider to make it not a separate step but an always available item in the other steps, that's disabled by default and enabled as soon as enough data is added to produce a valid annoation.
    - Add 'copy annotation url' button.
    - Facilitate and suggest local storage of resulting annotation
    - Suggest tools to open annotation (or XYZ) in: placemark, geojsonio, leaflet/openlayers/maplibre, qgis
- Facilitate the quality of GCPs and mask
  - Georeference
    - Improve the procedure: when specifying GCPs, first show resource and geo side to side. As soon as enough points have been added for the selected transformation algorithm, show them overlapping by default, alternating the transparency to show resource more apparent when collecting a resource coordinate and geo more apparent when collecting a geo coordinate. Update the transformation after every new point. Add option to go back to side by side view. Use different colors for points on resource or geo to make it clear which are which when both are displayed.
    - Allow loading gcps from elsewhere (using 'snap to' or 'select from')
      - Existing annotation (from file, url or your curator profile) ('select what you want to load: annotation, mask, transformation, projections, ...')
      - Geojson (selecting only points)
      - Shapefile (?) (selecting only points)
      - Other georeferening formats: QGIS, MapAnalyst, ESRI, ...
    - Allow manual input for both gcps and maks points: when listening for a click to specify a point, show a text explaining to click or to enter the coordinates (resource or geo) manually. This way it's possible to e.g. click for the resource coordinates of a mask, but specify the exact geo coordinates (which could be available in the appropriate crs) manually.
    - Allow speedy input of gcps when in 'side by side' mode: enough points have been added (for the specified transformation) add an option to suggest gcps on geo or on resource. Once selected the used species points on resource or geo, and is suggested to corresponding point on geo or resource respectively. The 'suggested' status is visualised using e.g. a color or transparency and encoded e.g. as `status = proposed`. This way the used can quickly add points e.g. on crossroads or trance a mask, and afterwards only has to check and slightly correct them.
  - Mask
    - Allow loading masks from elsewhere
      - Existing annotations
      - Geojson (selecting only polygons)
      - Shapefile (?) (selecting only polygons)
      - Svg
    - Button to generate mask gcps. This is useful when gcp's are the (four or more) corner points of the mask. This could be done by computing the bbox of the gpcs, or creating a polygon using the order in which they were specified.
    - Snap to points/lines: load an annotation or file (in geo or resource coordinates) and enable snappint to its points or lines when entering the points
    - Enable drawing straight lines (e.g. using shift) and straight angles.
    - Allow to (select and) move or scale gcps and mask. This can be useful when when loading gcps from an annotation with slightly different origin or resolution. See [this QGIS extension](https://gvellut.github.io/FreehandRasterGeoreferencer/).
    - Improve dragging of points. Currenlty this is buggy.
    - Use different color for resource and geo points. Indicate incomplete points (e.g. resource coordinates specified but geo coordinats missing) using color.
- Assess the quality of GCPs and mask
  - Add a way to asses the quality of the current annotation. This can also be used to scan the database and filter out pottentially bad annotations.
  - Warn for: outlier points. Use Principle Component Analysis to check the quality of points (wrt a Helmert or polynomial transform). This could enable to add a warning for outlier points that are possibly errors.
  - Warn for: gcp outside of mask
  - Warn for: large shear component on test-polynomial1 transformation
  - Warn for: large r^2 on test-Helmert transformation
  - Error for: self-intersections
  - Error for: failing test-triangulation (without gridpoints), to check for any other possible triangluation errors in the future
  - Error for: incomplete gcps
- Automation
  - Automated mask detection
  - Automated gcps

# Viewer

- Improve UI
  - Consider less visible Annotation (e.g. on the side, only first 10 characters visible, clear 'copy' button)
  - Consider giving input fields for annotation url and iiif urls different colors, and hinting wit `manifest url ` and `annotation url ` and adding a `(?)` buttonn that says `what is this?` on hover.
  - When zooming out, make large border or buffer around map to hint it's location. Add 'zoom to map'. Use case: e.g. 3 maps in Brussels and 2 maps in NY.
  - List export/share options for each map (see also editor): download annotation, manifest, geojson, gpcs, mask, download html of a page with with map, open annotation in qgis, leaflet, autocad, illustrator, and 'more' that links to docs with tutorials for converting to geotiff, ...
- Add distortion analysis
- Improve speed
  - Switch to MapLibre
  - Switch 'Image' tab to a straight transform on MapLibre, with continuous transition
  - Show 'loading' to user when IMAGEINFONEEDED or TILENEEDED, or untill FIRSTMAPTILELOADED are dispached.
- Improve WebGL
  - Clean WebGL memory better.
  - Remove seams
- Harmonise warped map options
  - Make a type that contains all options a warpedmap can have. These can then be stored centrally in the WarpedMap class, passed in a structured way to the webmap libraries (Leaflet, ...) and serialised in the URL (e.g. to load with a specific transformation, different then the one in the annotation).
  - WarpedMapOptions
    - TransformationType
    - Background Removal: 'Detect background color' and 'Remove color' (currently is WebGL2WarpedMap RenderOptions)
    - Colorise (currently in WebGL2WarpedMap RenderOptions)
    - Distortion analysis options
    - Opacity
    - Visible
    - UseMask
    - Z-index
    - Triangle size?
  - RenderOptions
    - All options from WarpedMapOptions, which can be set globally.
    - Other options?
- Improve Textures
  - Add a buffer around projectedGeoBbox when loading tiles, such as to load tiles that could become visible when panning. (Simpeler dan de resourceViewportRing te bufferen is mss gewoon het resultaat van de requestedTiles nemen en er telkens wat bij doen aan de zijkanten)
  - Improve texSubImage2D, which is the main timeconsuming step when plotting (many) tile. Replace with Texture Arrays
  - Consider using mapbox [shelf-pack](https://github.com/mapbox/shelf-pack) instead of potpack when storing tiles. 
- Improve TileCache
  - Improve tile selection and tilecache using pyramids
  - Consider setting a delay on tile remove (if tilecache size permits it)
- Improve developper experience
  - Improve variable names of stores to match the glossary and naming conventions. E.g. geoBbox, Map vs GeoreferencedMap, ...

# Webmap libraries: OL, Leaflet, MapLibre

- All
  - Allign threir three API's using 'implements'
- OpenLayers
  - Remove WebMapSource
- Leaflet
  - Document WarpedMapLayerOptions in Readme (e.g. to specify 'pane')
  - Submit to Leaflet Plugins page

# TileServer

- Improve usability
  - Support annotationPages
  - Support wider (global) EenderOptions
  - Include WMS and WMTS endpoints
- Improve caching and getting ImageInfo.
  - Refactor to make it work with Render, using some of Render's functions or implementing a specific Renderer (so it can use all of it's features).
    - Explore if WebGL2Renderer could be used
    - Explore if CanvasRenderer could be used
    - Actions to take if smaller refactor
      - Make it work with stdlib
      - Import transformations from main types package or transformation package
      - In `tilejson.rs`: bbox calculation: remove bbox polygon wrapping
      - Go throught createWarpedTileResponse() and adapt to new things from render refactor

# Arcade

- Include a 'Hard mode' using one or more of the following: only show inside of mask, disable viewing the map again, blur map to text can't be read, remove background
- Make it easy to make an arcade for a specific area or map collection

# QGIS Plugin

- Make QGIS Plugin using WebGL2Renderer

# CLI

- Implement loading other gcps and masks (see above)

# Developer tools

- Improve (remote) tests
  - Include local IIIF test server
  - Include local protomaps
  - Implement playwright to test build and compare webgl results

**Comment by itspangler** [comment]
Just flagging that last week, I opened a bunch of issues in the Editor repo without realizing that this may be the better place to put them: 

- Feature suggestion: https://github.com/allmaps/editor/issues/49
- Feature suggestion: https://github.com/allmaps/editor/issues/50
- Error/bug: https://github.com/allmaps/editor/issues/51

I suspect the error issue still belongs in the Editor repo, but let me know if I should migrate the other two feature requests/discussions to this post.
